python/src/main/java/org/apache/zeppelin/python/IPythonClient.java

```diff
@@ -131,11 +132,18 @@ public void onNext(ExecuteResponse executeResponse) {
       @Override
       public void onError(Throwable throwable) {
         try {
+          interpreterOutput.getInterpreterOutput().write(ExceptionUtils.getStackTrace(throwable));
           interpreterOutput.getInterpreterOutput().flush();
         } catch (IOException e) {
           LOGGER.error("Unexpected IOException", e);
         }
         LOGGER.error("Fail to call IPython grpc", throwable);
+        finalResponseBuilder.setStatus(ExecuteStatus.ERROR);
+
+        completedFlag.set(true);
+        synchronized (completedFlag) {
+          completedFlag.notify();
+        }
```

Comment:
that's great! just tested and glad to see errors are now getting its way to the fronted, without making spark interpreter appear "hanging".

Reply:
Thanks @Tagar for verifying and review.